### DISCUSSION TITLE
Pattern matching fails on types with implicit conversion to string

### DISCUSSION THREAD

**Discussion by AArnott** [dotnet/csharplang#7133] [Q&A] [answered]
I noticed today that pattern matching `switch` statements do not compile if the patterned type is implicitly convertible to string. Is this a C# language spec issue, or a bug in the roslyn compiler?

[Full repro](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQBMIGMD2A7DCMKOAGxgBoYQBLUgHwAEAmAAgFlWBvAWAChWg1owDMwgIwA2VjQC2AB1I0sNGKxzyICWDgQSADAApOANyik4EAJSsAvAD5WAIicBufkNYeho1sBw4pKwACtwA5hAwrsiRrgC+3oKJwgAsrABCxqyyNrwCnkLIAO6qWAAWrIY53MkFQlhQMRzi3CEgrDAIlqxxrO3ASFAA1u75dZ4NTexs/YMjtUIJYz38S/wszX0c3Esb01ucXHFAA===)

```cs
class M {
    public static implicit operator string(M value) => "";
    
    public bool P {get;set;}
    
    void B(M m) {
        switch (m) {
            case M1 { P: true } : break;
            case M2 : break;
        }
    }
}

class M1 : M {}
class M2 : M {}
```

> error CS8121: An expression of type 'string' cannot be handled by a pattern of type 'M1'.
error CS8121: An expression of type 'string' cannot be handled by a pattern of type 'M2'.

**Comment by HaloFour** [comment, marked as answer]
This feels like it might be legacy `switch` baggage, in that it used to allow expressions to be used in `switch` statements that were implicitly converted to `string` before pattern matching existed?
